onnxruntime/core/session/inference_session.cc

```diff
@@ -1505,6 +1505,134 @@ common::Status InferenceSession::ValidateOutputs(const std::vector<std::string>&
   return common::Status::OK();
 }
 
+#ifdef ENABLE_TRAINING
+std::pair<size_t, size_t> InferenceSession::GetBreakpointAndEndPoint() {
```

Comment:
Can this method move to `TrainingAgent`? I think it should be possible since `TrainingAgent` has access to `GetSessionState()`.

(Side-note: At some point, we may need to generalize this to support multiple yields from the same graph.)

Reply:
We can generalize this now, just wasn't sure if there was a use case for that. In anycase I plan to research ideas that can give us the execution plan (i.e start/end) by just looking at feeds and fetches, not doing it now since it won't add any perf value. There should not be need to insert a no-op like Yield to indicate this is where the execution needs to stop :)

---
In reply to: [613567371](https://github.com/microsoft/onnxruntime/pull/7324#discussion_r613567371) [](ancestors = 613567371)

Reply:
No need to generalize it in this PR. There is a use case, although it's not very urgent: basically we may want to yield individual gradient tensors one at a time, so that we can fire the autograd hooks for them sooner for computation/communication overlap - and potentially reclaim their memory sooner. 

@ashbhandare had prototyped a version of this with the old backend implementation, and it's on ice right now, but I think it will become a *lot* more efficient and effective with the changes in this PR!